shared/ssz/hash_cache.go

```diff
@@ -0,0 +1,279 @@
+package ssz
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"github.com/prysmaticlabs/prysm/shared/hashutil"
+	"github.com/prysmaticlabs/prysm/shared/params"
+	"k8s.io/client-go/tools/cache"
+)
+
+var (
+	// ErrNotMarakleRoot will be returned when a cache object is not a marakle root
+	ErrNotMarakleRoot = errors.New("object is not a markle root")
+	// maxCacheSize is 2x of the follow distance for additional cache padding.
+	// Requests should be only accessing blocks within recent blocks within the
+	// Eth1FollowDistance.
+	maxCacheSize = params.BeaconConfig().HashCacheSize
+
+	// Metrics
+	hashCacheMiss = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "powchain_hash_cache_miss",
+		Help: "The number of hash requests that aren't present in the cache.",
+	})
+	hashCacheHit = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "powchain_hash_cache_hit",
+		Help: "The number of hash requests that are present in the cache.",
+	})
+	hashCacheSize = promauto.NewGauge(prometheus.GaugeOpts{
+		Name: "powchain_hash_cache_size",
+		Help: "The number of hashes in the block cache",
+	})
+)
+
+// markleRoot specifies the hash of data in a struct
+type root struct {
+	Hash       common.Hash
+	MarkleRoot []byte
+}
+
+// hashKeyFn takes the  representation as the key for a hashInfo.
+func hashKeyFn(obj interface{}) (string, error) {
+	mRoot, ok := obj.(*root)
+	if !ok {
+		return "", ErrNotMarakleRoot
+	}
+
+	return mRoot.Hash.Hex(), nil
+}
+
+// newHashCache creates a new hash cache for storing/accessing root hashes from
+// memory.
+func newHashCache() *hashCacheS {
+	return &hashCacheS{
+		hashCache: cache.NewFIFO(hashKeyFn),
+	}
+}
+
+// hashCacheS struct with one queue for looking up by hash.
+type hashCacheS struct {
+	hashCache *cache.FIFO
+}
+
+// RootByHash fetches Root by its hash. Returns true with a
+// reference to the root if exists. Otherwise returns false, nil.
+func (b *hashCacheS) RootByHash(h common.Hash) (bool, *root, error) {
```

Comment:
So how would this work ? if you retrieve the root by its hash, you would have to generate the root in the first place which defeats the purpose of caching it

Reply:
Ok this is an important question. it retrieves the root by the hash of encoded form of the object. While the regular flow is to hash each branch of the object by itself and then hash them all into one

Reply:
In that case could you update the method decription to describe more accurately what the method does. When you say `Root by its hash` it reads as we hash the root of the encoded object, not that we hash the encoded object.

Reply:
fixed